gremlin/traversal/group.go

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy ofthe License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specificlanguage governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package traversal
+
+import (
+	"encoding/json"
+	"errors"
+	"strconv"
+
+	"github.com/skydive-project/skydive/flow"
+	"github.com/skydive-project/skydive/graffiti/graph/traversal"
+)
+
+// GroupTraversalExtension describes a new extension to enhance the topology
+type GroupTraversalExtension struct {
+	MoreThanToken traversal.Token
+}
+
+// GroupGremlinTraversalStep rawpackets step
+type GroupGremlinTraversalStep struct {
+	traversal.GremlinTraversalContext
+}
+
+// GroupTraversalStep rawpackets step
+type GroupTraversalStep struct {
+	GraphTraversal *traversal.GraphTraversal
+	flowGroupSet   map[string][]*flow.Flow
+	error          error
+}
+
+// NewGroupTraversalExtension returns a new graph traversal extension
+func NewGroupTraversalExtension() *GroupTraversalExtension {
+	return &GroupTraversalExtension{
+		MoreThanToken: traversalMoreThanToken,
+	}
+}
+
+// ScanIdent returns an associated graph token
+func (e *GroupTraversalExtension) ScanIdent(s string) (traversal.Token, bool) {
+	switch s {
+	case "MORETHAN":
+		return e.MoreThanToken, true
+	}
+	return traversal.IDENT, false
+}
+
+// ParseStep parse metrics step
+func (e *GroupTraversalExtension) ParseStep(t traversal.Token, p traversal.GremlinTraversalContext) (traversal.GremlinTraversalStep, error) {
+	switch t {
+	case e.MoreThanToken:
+		return &GroupGremlinTraversalStep{GremlinTraversalContext: p}, nil
+	}
+	return nil, nil
+}
+
+// Exec Group step
+func (g *GroupGremlinTraversalStep) Exec(last traversal.GraphTraversalStep) (traversal.GraphTraversalStep, error) {
+	switch last.(type) {
+	case *FlowTraversalStep:
+		fs := last.(*FlowTraversalStep)
+		return fs.Group(g.StepContext), nil
+	case *GroupTraversalStep:
+		group := last.(*GroupTraversalStep)
+		return group.MoreThan(g.StepContext), nil
```

Comment:
MoreThan should be an extension otherwise we don't know that the user called MoreThan, here we only know that thre previous step was group. If later we add more step after group we won't know what step to process. Aggregates is a good example